pkg/kubelet/nodeshutdown/systemd/inhibit_linux.go

```diff
@@ -0,0 +1,183 @@
+	duration := time.Duration(delay) * time.Microsecond
+	return duration, nil
+}
+
+// InhibitShutdown creates an systemd inhibitor by calling logind's Inhibt() and returns the inhibitor lock
+// see https://www.freedesktop.org/wiki/Software/systemd/inhibit/ for more details.
+func (bus *DBusCon) InhibitShutdown() (InhibitLock, error) {
+	obj := bus.SystemBus.Object(logindService, logindObject)
+	what := "shutdown"
+	who := "kubelet"
+	why := "Kubelet needs time to handle node shutdown"
+	mode := "delay"
+
+	call := obj.Call("org.freedesktop.login1.Manager.Inhibit", 0, what, who, why, mode)
+	if call.Err != nil {
+		return InhibitLock(0), fmt.Errorf("failed creating systemd inhibitor: %v", call.Err)
+	}
+
+	var fd uint32
+	err := call.Store(&fd)
+	if err != nil {
+		return InhibitLock(0), fmt.Errorf("failed storing inhibit lock file descriptor: %v", err)
+	}
+
+	return InhibitLock(fd), nil
+}
+
+// ReleaseInhibitLock will release the underlying inhibit lock which will cause the shutdown to start.
+func (bus *DBusCon) ReleaseInhibitLock(lock InhibitLock) error {
+	err := syscall.Close(int(lock))
+
+	if err != nil {
+		return fmt.Errorf("unable to close systemd inhibitor lock: %v", err)
+	}
+
+	return nil
+}
+
+// ReloadLogindConf uses dbus to send a SIGHUP to the systemd-logind service causing logind to reload it's configuration.
+func (bus *DBusCon) ReloadLogindConf() error {
+	systemdService := "org.freedesktop.systemd1"
+	systemdObject := "/org/freedesktop/systemd1"
+	systemdInterface := "org.freedesktop.systemd1.Manager"
+
+	obj := bus.SystemBus.Object(systemdService, dbus.ObjectPath(systemdObject))
+	unit := "systemd-logind.service"
+	who := "all"
+	var signal int32 = 1 // SIGHUP
+
+	call := obj.Call(systemdInterface+".KillUnit", 0, unit, who, signal)
+	if call.Err != nil {
+		return fmt.Errorf("unable to reload logind conf: %v", call.Err)
+	}
+
+	return nil
+}
+
+// MonitorShutdown detects the a node shutdown by watching for "PrepareForShutdown" logind events.
+// see https://www.freedesktop.org/wiki/Software/systemd/inhibit/ for more details.
+func (bus *DBusCon) MonitorShutdown() (<-chan bool, error) {
+	err := bus.SystemBus.AddMatchSignal(dbus.WithMatchInterface(logindInterface), dbus.WithMatchMember("PrepareForShutdown"), dbus.WithMatchObjectPath("/org/freedesktop/login1"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	busChan := make(chan *dbus.Signal, 1)
+	bus.SystemBus.Signal(busChan)
+
+	shutdownChan := make(chan bool, 1)
+
+	go func() {
+		for {
+			select {
+			case event := <-busChan:
+				if event == nil || len(event.Body) == 0 {
+					klog.Errorf("Failed obtaining shutdown event, PrepareForShutdown event was empty")
+				}
+				shutdownActive, ok := event.Body[0].(bool)
+				if !ok {
+					klog.Errorf("Failed obtaining shutdown event, PrepareForShutdown event was not bool type as expected")
+					return
+				}
+				klog.V(1).Infof("Received node shutdown event; shutdown active: %t", shutdownActive)
```

Comment:
There will be this log and one in the manager with the very similar content. Are both needed?

Reply:
good point, removed this since the manager will print it anyways.